tests/queries/0_stateless/02403_date_time_narrowing.sql

```diff
@@ -1,51 +1,51 @@
 SELECT toDate(toDate32('1930-01-01', 'UTC'), 'UTC'),
-       toDate(toDate32('2150-01-01', 'UTC'), 'UTC');
+       toDate(toDate32('2151-01-01', 'UTC'), 'UTC');
 
 SELECT toDate(toDateTime64('1930-01-01 12:12:12.12', 3, 'UTC'), 'UTC'),
-       toDate(toDateTime64('2150-01-01 12:12:12.12', 3, 'UTC'), 'UTC');
+       toDate(toDateTime64('2151-01-01 12:12:12.12', 3, 'UTC'), 'UTC');
 
 SELECT toDateTime(toDateTime64('1930-01-01 12:12:12.12', 3, 'UTC'), 'UTC'),
-       toDateTime(toDateTime64('2150-01-01 12:12:12.12', 3, 'UTC'), 'UTC');
+       toDateTime(toDateTime64('2151-01-01 12:12:12.12', 3, 'UTC'), 'UTC');
 
 SELECT toDateTime(toDate32('1930-01-01', 'UTC'), 'UTC'),
-       toDateTime(toDate32('2150-01-01', 'UTC'), 'UTC');
+       toDateTime(toDate32('2151-01-01', 'UTC'), 'UTC');
 
-SELECT toDateTime(toDate('2140-01-01', 'UTC'), 'UTC');
+SELECT toDateTime(toDate('2141-01-01', 'UTC'), 'UTC');
 
 SELECT 'toStartOfDay';
-SELECT toStartOfDay(toDate('2141-01-01', 'UTC'), 'UTC'),
-       toStartOfDay(toDate('1941-01-01', 'UTC'), 'UTC'),
+SELECT toStartOfDay(toDate('1930-01-01', 'UTC'), 'UTC'),
+       toStartOfDay(toDate('2141-01-01', 'UTC'), 'UTC'),
+       toStartOfDay(toDate32('1930-01-01', 'UTC'), 'UTC'),
        toStartOfDay(toDate32('2141-01-01', 'UTC'), 'UTC'),
-       toStartOfDay(toDate32('1941-01-01', 'UTC'), 'UTC'),
-       toStartOfDay(toDateTime64('1941-01-01 12:12:12.123', 3, 'UTC'), 'UTC'),
+       toStartOfDay(toDateTime64('1930-01-01 12:12:12.123', 3, 'UTC'), 'UTC'),
        toStartOfDay(toDateTime64('2141-01-01 12:12:12.123', 3, 'UTC'), 'UTC');
 
 SELECT 'toMonday';
-SELECT toMonday(toDate32('2151-01-01', 'UTC')),
-       toMonday(toDate32('1941-01-01', 'UTC')),
-       toMonday(toDateTime64('1941-01-01 12:12:12.123', 3, 'UTC'), 'UTC'),
+SELECT toMonday(toDate32('1930-01-01', 'UTC')),
+       toMonday(toDate32('2151-01-01', 'UTC')),
+       toMonday(toDateTime64('1930-01-01 12:12:12.123', 3, 'UTC'), 'UTC'),
        toMonday(toDateTime64('2151-01-01 12:12:12.123', 3, 'UTC'), 'UTC');
 
 SELECT 'toStartOfMonth';
-SELECT toStartOfMonth(toDate32('2151-01-01', 'UTC')),
-       toStartOfMonth(toDate32('1941-01-01', 'UTC')),
-       toStartOfMonth(toDateTime64('1941-01-01 12:12:12.123', 3, 'UTC'), 'UTC'),
+SELECT toStartOfMonth(toDate32('1930-01-01', 'UTC')),
+       toStartOfMonth(toDate32('2151-01-01', 'UTC')),
+       toStartOfMonth(toDateTime64('1930-01-01 12:12:12.123', 3, 'UTC'), 'UTC'),
        toStartOfMonth(toDateTime64('2151-01-01 12:12:12.123', 3, 'UTC'), 'UTC');
 
 SELECT 'toLastDayOfMonth';
-SELECT toLastDayOfMonth(toDate32('2151-01-01', 'UTC')),
-       toLastDayOfMonth(toDate32('1941-01-01', 'UTC')),
-       toLastDayOfMonth(toDateTime64('1941-01-01 12:12:12.123', 3, 'UTC'), 'UTC'),
+SELECT toLastDayOfMonth(toDate32('1930-01-01', 'UTC')),
```

Comment:
For unclear reasons, function `toLastDayOfMonth` is not documented in the official docs (but it [should be](https://github.com/ClickHouse/ClickHouse/pull/34394/files)).

Reply:
Function `toLastDayOfMonth` has the interesting property that when called on 1th, 2nd, 3rd, 4th, 5th, 6th June 2148, it should (in theory) produce 30th June 2148 but that is outside the range of the date type. So truncation happens not only (potentially) for inputs but also for outputs. Note that all other tested functions (`toStartOfXYZ`) don't have a similar property for their output values because date types `Date`, `Date32`, `DateTime` and `DateTime64` all start on 1st January. 

So maybe it's worth to test that special case as well.

EDIT1: `toMonday` is also interesting. As far as I see, it produces a 16 bit value, i.e. data type `Date`. The computed "monday" for input values 1th, 2th, 3th and 4th January 1970 as inputs should (in theory) be 29th December 1969 which is outside the range of `Date`. Surely also worth a test. (--> https://www.timeanddate.com/calendar/?year=1969&country=8)

EDIT2: Reading the [docs](https://clickhouse.com/docs/en/sql-reference/functions/date-time-functions/#tostartofweektmode), method `toStartOfWeek`potentially has the same problem. It is not tested here.

Reply:
```
┌─toLastDayOfMonth(toDate('2149-06-03'))─┐
│                             1970-01-24 │
└────────────────────────────────────────┘
```

Reply:
At least the latter two work:
```
┌─toMonday(toDate('1970-01-02'))─┐
│                     1970-01-01 │
└────────────────────────────────┘

┌─toStartOfWeek(toDate('1970-01-02'))─┐
│                          1970-01-01 │
└─────────────────────────────────────┘
```